File: data-structures/queue/queue_test.go
package queue

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestQueue(t *testing.T) {
	t.Run("full queue / dequeue", func(t *testing.T) {
		q := NewQueue[int](5)
		assert.True(t, q.IsEmpty())
		for i := range 5 {
			assert.True(t, q.Enqueue(i+1))
		}

		assert.Equal(t, 1, *q.Peek())
		assert.True(t, q.Size() == 5)
		assert.True(t, q.IsFull())

		isEnqueued := q.Enqueue(10)
		assert.True(t, !isEnqueued)

		for i := range 5 {
			v := q.Dequeue()
			assert.NotNil(t, v)
			assert.Equal(t, i+1, *v)
		}
		assert.True(t, q.IsEmpty())
		assert.True(t, !q.IsFull())

		v := q.Dequeue()
		assert.Nil(t, v)
	})

	t.Run("partial queue / dequeue", func(t *testing.T) {
		q := NewQueue[int](3)
		assert.Nil(t, q.Peek())
		assert.True(t, q.Enqueue(10))
		assert.True(t, q.Enqueue(20))
		assert.NotNil(t, q.Dequeue())
		assert.True(t, q.Enqueue(30))
		assert.True(t, q.Enqueue(40))
		assert.True(t, q.IsFull())
		assert.True(t, !q.Enqueue(50))

		assert.NotNil(t, q.Dequeue())
		assert.True(t, q.Enqueue(60))
		assert.True(t, q.IsFull())
	})
}

File: app/data_structures/queue/queue_test.py
from unittest import TestCase
from .queue import Queue


class TestQueue(TestCase):
    def test_normal_usage(self) -> None:
        q = Queue[int](capacity=5)
        self.assertTrue(q.is_empty())
        for i in range(4):
            q.enqueue(i + 1)

        self.assertEqual(len(q), 4)
        self.assertEqual(q.capacity, 5)

        all_elements = q.__iter__()
        self.assertEqual(list(all_elements), [1, 2, 3, 4])

        first = q.dequeue()
        self.assertEqual(first, 1)
        self.assertEqual(len(q), 3)

        q.enqueue(5)
        q.enqueue(6)
        self.assertEqual(len(q), 5)
        self.assertEqual(q.capacity, 5)
        self.assertTrue(q.is_full())

        q.enqueue(7)
        self.assertEqual(len(q), 6)
        self.assertEqual(q.capacity, 10)

        peeked = q.peek()
        self.assertEqual(peeked, 2)
        self.assertEqual(len(q), 6)

    def test_empty_list(self) -> None:
        q = Queue[int](capacity=5)
        items = q.__iter__()
        self.assertEqual(list(items), [])
        self.assertIsNone(q.peek())
        self.assertIsNone(q.dequeue())
